file_scanning/clamd/clamd.go

```diff
@@ -0,0 +1,229 @@
+package clamd
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+)
+
+const (
+	RES_OK          = "OK"
+	RES_FOUND       = "FOUND"
+	RES_ERROR       = "ERROR"
+	RES_PARSE_ERROR = "PARSE ERROR"
+)
+
+type Clamd struct {
+	address string
+}
+
+type Stats struct {
+	Pools    string
+	State    string
+	Threads  string
+	Memstats string
+	Queue    string
+}
+
+type ScanResult struct {
+	Raw         string
+	Description string
+	Path        string
+	Hash        string
+	Size        int
+	Status      string
+}
+
+var EICAR = []byte(`X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*`)
+
+func (c *Clamd) newConnection() (conn *CLAMDConn, err error) {
+
+	var u *url.URL
+
+	if u, err = url.Parse(c.address); err != nil {
+		return
+	}
+
+	conn, err = newCLAMDTcpConn(u.Host)
+
+	return
+}
+
+func (c *Clamd) simpleCommand(command string) (chan *ScanResult, error) {
+	conn, err := c.newConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	err = conn.sendCommand(command)
+	if err != nil {
+		return nil, err
+	}
+
+	ch, wg, err := conn.readResponse()
+
+	go func() {
+		wg.Wait()
+		conn.Close()
+	}()
+
+	return ch, err
+}
+
+/*
+Check the daemon's state (should reply with PONG).
+*/
+func (c *Clamd) Ping() error {
+	ch, err := c.simpleCommand("PING")
+	if err != nil {
+		return err
+	}
+
+	select {
+	case s := (<-ch):
+		switch s.Raw {
+		case "PONG":
+			return nil
+		default:
+			return errors.New(fmt.Sprintf("Invalid response, got %s.", s.Status))
+		}
+	}
+
+	return nil
+}
+
+/*
+Print program and database versions.
+*/
+func (c *Clamd) Version() (chan *ScanResult, error) {
+	dataArrays, err := c.simpleCommand("VERSION")
+	return dataArrays, err
+}
+
+/*
+On this command clamd provides statistics about the scan queue, contents of scan
+queue, and memory usage. The exact reply format is subject to changes in future
+releases.
+*/
+func (c *Clamd) Stats() (*Stats, error) {
+	ch, err := c.simpleCommand("STATS")
+	if err != nil {
+		return nil, err
+	}
+
+	stats := &Stats{}
+
+	for s := range ch {
+		if strings.HasPrefix(s.Raw, "POOLS") {
+			stats.Pools = strings.Trim(s.Raw[6:], " ")
+		} else if strings.HasPrefix(s.Raw, "STATE") {
+			stats.State = s.Raw
+		} else if strings.HasPrefix(s.Raw, "THREADS") {
+			stats.Threads = s.Raw
+		} else if strings.HasPrefix(s.Raw, "QUEUE") {
+			stats.Queue = s.Raw
+		} else if strings.HasPrefix(s.Raw, "MEMSTATS") {
+			stats.Memstats = s.Raw
+		} else if strings.HasPrefix(s.Raw, "END") {
+		} else {
+			return nil, errors.New(fmt.Sprintf("Unknown response, got %s.", s))
+		}
```

Comment:
This could be written as a simple `switch` statement
```
switch {
case strings.HasPrefix(s.Raw, "POOLS"):
	stats.Pools = strings.Trim(s.Raw[6:], " ")
case strings.HasPrefix(s.Raw, "STATE"):
	stats.State = s.Raw
case strings.HasPrefix(s.Raw, "THREADS"):
	stats.Threads = s.Raw
case strings.HasPrefix(s.Raw, "MEMSTATS"):
	stats.Memstats = s.Raw
case strings.HasPrefix(s.Raw, "END"):
default:
	return nil, fmt.Errorf("invalid response, got %v", s)
}
```

Also change `errors.New()` to just use `fmt.Errorf` as that function just returns an `error` like you're expecting. Also based on the Go handbook, error's should not have capitalisation to start unless beginning with proper nouns or acronyms also they shouldn't end in punctuation